### DISCUSSION TITLE
How to export to a different language? Writing a new exporter.

### DISCUSSION THREAD

**Discussion by akiross** [tweag/nickel#1481] [Q&A]
Hello!

While YAML, TOML and JSON are pretty diffused, there are a lot of other languages that could be target for Nickel (e.g. nix, hcl).

I was wondering if there is any documentation/examples on how to extend nickel to export to a new language (or, in general, how to synthesize a custom file from a nickel source).

Thank you

**Comment by akiross** [comment]
Found this relevant discussion: https://github.com/tweag/nickel/discussions/665

**Reply by yannham** [reply]
Right, I think the information in #665 is still up to date. In order to provide a generic way of writing new importers, we have to think about a kind of adds-on/plugin system for the interpreter first. Do you have a particular target format in mind, or were you just wondering out of curiosity? Note that for Nix and HCL (and maybe others), you would probably export to a subset of those languages which is more or less JSON - in that case, it might be easier to just export to JSON and import the result from those languages (that's what [nickel nix](https://github.com/nickel-lang/nickel-nix) and [terraform-nickel](https://github.com/tweag/tf-ncl) are doing, respectively for Nix and Terraform precisely)

**Reply by akiross** [reply]
It's mostly curiosity, but there are actually a couple of concrete use cases where I could apply nickel. One is for an internal R&D project that involves SQL files and Dockerfiles, but the two are somewhat interleaved, so there's a bunch of factors that influence what goes in the Dockerfiles and, consequently, in the SQL files.

I can actually use other solutions (e.g. cookiecutter, python scripts, AWK), but I wanted to consider nickel.

Just for the sake of learning, I'll try to use the `--raw` formatting using strings and try to build a custom exporter.
